Feat/create new table dialog box

signekb opened the pull request:
## Description

<!-- 
This template covers all PRs for Seedcase, please note that if you are
submitting a PR for changes to:

a) General documentation you should delete the sections Testing, Code
Documentation, and the first part of the Author Checklist.

b) Code you should delete the second section of the Author Checklist.

Otherwise, delete any sections that don't apply.
-->

<!-- DO NOT LEAVE THIS SECTION BLANK -->

- This PR adds a dialog overlay on the data import page with two fields: `name` and `description`
- The `name` field includes two validators:
    - A validator that checks that the table name doesn't include special characters (only a-z, A-Z, 0-9, hyphen (-), and/or underscore (_)
    - A validator that checks that a table with the same name doesn't exist already
- If table `name` and `description` passes validation, the information is saved in the `TableMetadata`
-  I have included tests for the validators, the form, and the data import view
- In addition, this PR adds a (?) button on the data import page that opens a dialog where we can write a helper message and a link to our documentation. (this needs to be updated later ofc)

I have added a detailed description of the changes in the bottom of this PR to (hopefully) ease the review process, if something is unclear. 

## Related Issues

<!-- List issues the PR closes -->

Closes #75, #76 and #125

## Testing

- [X] Yes


<!-- Please explain why the tests are not needed for this PR here -->

## Reviewer Focus
 
<!-- Any particular section the reviewer should focus on, anywhere that would be a good place to start? -->

- Anything in the code that's unclear? / Suggestions for better ways of doing this? 
- Variable naming and doc string descriptions
- Tests: Enough / too many / some that aren't necessary?

## Checklist

<!-- This is to help you determine if your work is ready to be reviewed, if an item is not relevant then you can mark it as done (because you have checked and found that it isn't needed) -->

For all PRs that are not general documentation

- [X] Tests accompany or reflect changes to the code
- [X] Tests ran and passed locally
- [X] Ran the linter and formatter (NB: But only on the files relevant to this PR)
- [X] Build has passed locally
- [X] Relevant documentation has been updated

### Overview of changes
Here's an overview of the changes in the order they appear in the directory (I hope this will ease the review process): 

- **app/forms.py**: 
    - I have created a `forms.py` file with a [ModelForm](https://docs.djangoproject.com/en/5.0/topics/forms/modelforms/) called `TableMetadataForm` . This class uses the model `TabelMetadata` and the fields `name` and `description` from that model. It's these fields that we want users to fill out in the dialog overlay when they create a table (as shown in the wireframes on Figma). 
    - Within this form, I call the two validators, I have created in `validators.py`
- **app/templates/data-import.html**:
    -  I have created a "create-table-dialog" that opens, when a user clicks "create new table". This dialog includes two validations, which will be described below.
    - In addition, I have added a help button that opens the "help-dialog" when clicked.
- **app/validators/field_validators.py**:
    - `validate_no_special_characters`: A regex validator that fails, if the field includes special characters. We can reuse this validators for other fields where we don't want special characters
    - `validate_table_name_does_not_exist`: A validator that checks that a table name doesn't exist in the database already.
- **app/views.py**
    - I have updated the `data_import` view to include the `TableMetadataForm`. When the dialog opens, the request will be `GET` and a blank form is created. When the user clicks "Create", the request changes to `POST` (see `data-import.html`) and the form's content is checked by the validators. If the validation fails, the error message from the relevant validator is shown. If the validation succeeds, it's saved and the url is redirected to `file-upload/{table_id}`, where {table-id} is the id of the table that was just created. This is useful for the file-upload page, so this page can get the table name from the id.
- **app/tests/tests_forms.py**: 
    - Tests of the form `TableMetadataForm`. 
- **app/tests/tests_validators.py**: 
    - Tests for the two validators `validate_no_special_characters` and `validate_table_name_does_not_exist`
- **app/tests/tests_views.py**: 
    - Tests for the view (does it render, redirect, etc.) when we expect it to

signekb committed:
feat: init/wip

signekb committed:
Merge branch 'main' into feat/create-new-table-dialog-box

signekb committed:
feat: change tablecreation form to modelform

this way, it's directly related to the tablemetadata model

signekb committed:
feat: update data_import view to use and validate tablemetadataform

this validation only includes whether a field is blank or not

signekb committed:
feat: hotfix - created_at defaults to none until we implement users

signekb committed:
feat: update html to utilise modelform fields

additionally, the dialog is renames to 'create-table-dialog' and the cancel button is made 'a' to avoid form validation

signekb committed:
feat: add table id to file-upload url

signekb committed:
style: remove blank line

signekb committed:
fix: remove default=None from created_by

not needed anymore with new implementation of form

signekb committed:
feat: add validator to prevent special characters in the table name field

signekb committed:
feat: add validator to prevent creating tables with name that already exists

signekb committed:
style: add and remove blank lines

signekb committed:
feat: show validation errors for name field

atm we have two possible validation errors: table name already exists in db and special characters

signekb committed:
feat: center align dialog header

signekb committed:
feat: add icon to align with wireframes

signekb committed:
fix: remove description text field - we use the form field now

signekb committed:
fix: remove data-ui from create button to avoid closing dialog when validation fails

signekb committed:
style: add comments

signekb committed:
feat: add help button with dialog overlay

signekb committed:
style: add whitespace before table dialog

signekb committed:
style: :lipstick: change hyperlink colour to blue

signekb committed:
feat: include val of table already existing directly in the modelform

signekb committed:
refactor: rename val of no special characters

signekb committed:
refactor: explicitly state that create is a submit button

signekb committed:
fix: add button to cancel class

By adding this, it looks identical to the "create" button (except for the text ofc)

signekb committed:
fix: if errors in form, show dialog

This way, the dialog does not close when validation fails

signekb committed:
style: simplify comments

signekb committed:
refactor: move val table already exists to form instead of model

signekb committed:
revert: black formatting

Not relevant for this PR (now that I haven't made other changes to this file)

signekb committed:
revert: remove irrelevant comment

signekb committed:
style: rename val_table_does_not_exist to include *not*

signekb committed:
feat: update special characters validator to include field_name in error msg

signekb committed:
refactor: use redirect instead of httpresponseredirect

this way, data_import only uses django shortcuts for HttpResponses (render and redirect, respectively)

signekb committed:
style: typehints, comments, and format

signekb committed:
refactor: move validation of special character to validator

signekb committed:
style: rename variable "name" to "name_value"

signekb committed:
feat: add error code to validate_table_name_does_not_exist

signekb committed:
refactor: add type hints for returns

signekb committed:
test: :sparkles: tablemetadataform validation tests

signekb committed:
test: :sparkles: add for validators special character and table exists

signekb committed:
docs: remove old/irrelevant code comments

signekb commented:
@lwjohnst86 @philter87, in the fly preview, it fails whenever the request goes to "POST" (I think). I'm not sure why this happens. Do you have an idea?

philter87 reviewed (dismissed):
Beautiful :) I have much to learn about making a clean PR! I think everything is very clear. 

I don't think testing the view is so important, because you have tested the validators and the form.

philter87 commented:
> @lwjohnst86 @philter87, in the fly preview, it fails whenever the request goes to "POST" (I think). I'm not sure why this happens. Do you have an idea?

I think this is weird too! I didn't do for my hobby project.

However, I think it should work if you add: 
CSRF_TRUSTED_ORIGINS = ["https://seedcase-sprout*.fly.dev"]

But I might need to investigate a bit more to know if it is secure

signekb committed:
docs: remove extra "upload"

signekb committed:
style: add whitespace

signekb committed:
feat: :sparkles: add tests for data_import view

signekb committed:
style: add empty line after docstrings

signekb committed:
Merge branch 'main' into feat/create-new-table-dialog-box

signekb commented on app/forms.py:
```diff
@@ -0,0 +1,24 @@
+        model = TableMetadata
+        fields = ["name", "description"]
+
+    def clean_name(self):
```
Great idea. Done! 🎉

signekb committed:
refactor: add typehints and docstring to clean_name

signekb committed:
refactor: :art: specify imports from django.forms

signekb commented:
I'm through my edits based on feedback from @philter87, and I have now added tests for the view

pchmia committed:
fix: trying to fix csrf

pchmia committed:
fix: trying to fix csrf adding comment

pchmia committed:
fix: trying to fix csrf adding comment

pchmia committed:
fix: missing file

pchmia committed:
fix: trying CSRF_TRUSTED_ORIGINS

philter87 commented:
I did a couple of things to fix the CSRF POST request issue:
- I added CSRF_TRUSTED_ORIGINS = [f"https://{APP_NAME}.fly.dev/"]
- I resolved the merge conflict, because it prevented deploys to fly.io (needed to validate the CSRF issue)

signekb commented on app/tests/tests_validators.py:
```diff
@@ -0,0 +1,112 @@
+        # Assert
+        self.assertEqual(
+            context.exception.message,
+            f"Please provide a {field_name} without special characters",
```
Uh, I like that! Will change it

signekb committed:
refactor: :lipstick: update error msg of no special characters validator

Now it's what to do (which characters to use), instead of what not to do (don't use special characters). Based on @K-Beicher 's feedback

signekb committed:
test: update tests with new error message for no_special_characters validator

philter87 commented on app/templates/data-import.html:
```diff
@@ -4,13 +4,11 @@
-            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 -960 960 960" ><path d="M440-440H200v-80h240v-240h80v240h240v80H520v240h-80v-240Z"/></svg>
-          </article>
-        </a>
+        <article class="card-button-size" data-ui="#create-table-dialog">
```
A small thing. The cursor is not turning into a hand when I hover the button - this disappeared when the <a> tag was removed. Add "cursor: pointer" to "card-button-size" or add style="cursor:pointer"

signekb replied:
Aha! Good catch! I will add that :hand:

signekb committed:
feat: add cursor pointer to card-button-size buttons

signekb committed:
Merge branch 'main' into feat/create-new-table-dialog-box

lwjohnst86 committed:
style: :art: Formatting and style of docstrings, ran ruff

lwjohnst86 committed:
style: :lipstick: Use edit icon, not mobile icon

The mobile friendly icon didn't seem appropriate, so changed it to edit icon.

lwjohnst86 committed:
test: :white_check_mark: Add a test that checks for an invalid form

lwjohnst86 reviewed (approved):
Looks so fantastic! Great job 😁 I only made some minor changes and added a test to check for invalid input.

signekb commented:
@lwjohnst86 
I believe you have created a duplicate test? Isn't the invalid input you added on L61-65 the same test as the one on L8-15?
https://github.com/seedcase-project/seedcase-sprout/blob/a9fe5ec8e1c73bc18376b29abdf6c4949a4933a7/app/tests/tests_forms.py#L8-L15

https://github.com/seedcase-project/seedcase-sprout/blob/a9fe5ec8e1c73bc18376b29abdf6c4949a4933a7/app/tests/tests_forms.py#L61-L65

lwjohnst86 commented:
Ah, you're right @signekb, oops!

signekb commented:
I guess that's the danger of skipping reviews ;)

lwjohnst86 commented:
😆 that's very true!
